### PR TITLE
Register default services directly in AddMvc and remove MvcServices

### PR DESCRIPTION
**Pull request by kichalla**
https://github.com/aspnet/Mvc/issues/1634

@rynowak @pranavkm @HaoK

### CODE REVIEW

**Comment by Eilon** [comment]
Are the only real changes here that it now call `TryAddXyz`?

Also per @HaoK's comments from #1634 are any of these multiply-registered services that should use `AddXyz` or are they all single-instance services? (Did that even make any sense? :smile: )

File: src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -16,7 +33,158 @@ public static class MvcServiceCollectionExtensions
         public static IServiceCollection AddMvc([NotNull] this IServiceCollection services)
         {
             ConfigureDefaultServices(services);
-            services.TryAdd(MvcServices.GetDefaultServices());
+
+            // Options and core services.
+            services.TryAdd(ServiceDescriptor.Transient<IConfigureOptions<MvcOptions>, MvcOptionsSetup>());
+            services.TryAdd(
```
Comment (HaoK) [inline src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs]: This two probably should be a normal Adds, as multiple configure options are allowed and are run in order (its an IEnueramble

File: src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs [hunk 2]
```diff
@@ -16,7 +33,158 @@ public static class MvcServiceCollectionExtensions
...
+            //
+            // The DefaultModelMetadataProvider does significant caching and should be a singleton.
+            services.TryAdd(ServiceDescriptor.Singleton<IModelMetadataProvider, DefaultModelMetadataProvider>());
+            services.TryAdd(ServiceDescriptor.Transient<ICompositeMetadataDetailsProvider>(serviceProvider =>
+            {
+                var options = serviceProvider.GetRequiredService<IOptions<MvcOptions>>().Options;
+                return new DefaultCompositeMetadataDetailsProvider(options.ModelMetadataDetailsProviders);
+            }));
+
+            services.TryAdd(ServiceDescriptor.Instance(typeof(JsonOutputFormatter), new JsonOutputFormatter()));
```
Comment (HaoK) [inline src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs]: Can this just be `Singleton<JsonOutputFormatter, JsonOutputFormatter>()`?

**Comment by HaoK** [comment]
Yeah so the general idea is, anything that is being resolved as IEnumerable of a service, you want don't want to use TryAdd, because well, you'll never get more than one.  So I noticed that issue for IConfigureOptions, I am not familiar enough with all the other MVC services, but that's the review that needs to be done...

**Comment by rynowak** [comment]
There's a few others that are multi-service. I'll take a look.

File: src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs [hunk 3]
```diff
@@ -16,7 +33,158 @@ public static class MvcServiceCollectionExtensions
         public static IServiceCollection AddMvc([NotNull] this IServiceCollection services)
         {
             ConfigureDefaultServices(services);
-            services.TryAdd(MvcServices.GetDefaultServices());
+
+            // Options and core services.
+            services.TryAdd(ServiceDescriptor.Transient<IConfigureOptions<MvcOptions>, MvcOptionsSetup>());
+            services.TryAdd(
+                ServiceDescriptor.Transient<IConfigureOptions<RazorViewEngineOptions>, RazorViewEngineOptionsSetup>());
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs]: Anything with options is a multi-service.

File: src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs [hunk 4]
```diff
@@ -16,7 +33,158 @@ public static class MvcServiceCollectionExtensions
...
+            services.TryAdd(ServiceDescriptor.Singleton<IControllerFactory, DefaultControllerFactory>());
+
+            services.TryAdd(ServiceDescriptor.Transient<IControllerActivator, DefaultControllerActivator>());
+
+            // This accesses per-request services
+            services.TryAdd(ServiceDescriptor.Transient<IActionInvokerFactory, ActionInvokerFactory>());
+
+            // This provider needs access to the per-request services, but might be used many times for a given
+            // request.
+            services.TryAdd(ServiceDescriptor.Transient<IActionConstraintProvider, DefaultActionConstraintProvider>());
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs]: multiservice

File: src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs [hunk 5]
```diff
@@ -16,7 +33,158 @@ public static class MvcServiceCollectionExtensions
...
+                .Transient<IControllerActionArgumentBinder, DefaultControllerActionArgumentBinder>());
+            services.TryAdd(ServiceDescriptor.Transient<IObjectModelValidator>(serviceProvider =>
+            {
+                var options = serviceProvider.GetRequiredService<IOptions<MvcOptions>>().Options;
+                var modelMetadataProvider = serviceProvider.GetRequiredService<IModelMetadataProvider>();
+                return new DefaultObjectValidator(options.ValidationExcludeFilters, modelMetadataProvider);
+            }));
+
+            services.TryAdd(ServiceDescriptor
+                .Transient<IActionDescriptorProvider, ControllerActionDescriptorProvider>());
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs]: multiservice

File: src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs [hunk 6]
```diff
@@ -16,7 +33,158 @@ public static class MvcServiceCollectionExtensions
...
+
+            services.TryAdd(ServiceDescriptor.Transient<IActionInvokerProvider, ControllerActionInvokerProvider>());
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs]: multiservice

File: src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs [hunk 7]
```diff
@@ -16,7 +33,158 @@ public static class MvcServiceCollectionExtensions
...
+
+            services.TryAdd(ServiceDescriptor
+                .Singleton<IActionDescriptorsCollectionProvider, DefaultActionDescriptorsCollectionProvider>());
+
+            // The IGlobalFilterProvider is used to build the action descriptors (likely once) and so should
+            // remain transient to avoid keeping it in memory.
+            services.TryAdd(ServiceDescriptor.Transient<IGlobalFilterProvider, DefaultGlobalFilterProvider>());
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs]: multiservice

File: src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs [hunk 8]
```diff
@@ -16,7 +33,158 @@ public static class MvcServiceCollectionExtensions
...
+            services.TryAdd(ServiceDescriptor.Transient<IFilterProvider, DefaultFilterProvider>());
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs]: multiservice

File: src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs [hunk 9]
```diff
@@ -16,7 +33,158 @@ public static class MvcServiceCollectionExtensions
...
+            // Security and Authorization
+            services.TryAdd(ServiceDescriptor.Singleton<IClaimUidExtractor, DefaultClaimUidExtractor>());
+            services.TryAdd(ServiceDescriptor.Singleton<AntiForgery, AntiForgery>());
+            services.TryAdd(ServiceDescriptor
+                .Singleton<IAntiForgeryAdditionalDataProvider, DefaultAntiForgeryAdditionalDataProvider>());
+
+            // Api Description
+            services.TryAdd(ServiceDescriptor
+                .Singleton<IApiDescriptionGroupCollectionProvider, ApiDescriptionGroupCollectionProvider>());
+            services.TryAdd(ServiceDescriptor.Transient<IApiDescriptionProvider, DefaultApiDescriptionProvider>());
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs]: multiservice

**Comment by rynowak** [comment]
Also take a look at the webapi compat shim project.

File: src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs (hunk 4 above)
Comment (kichalla) [inline src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs]: :+1:

File: src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs (hunk 2 above)
Comment (kichalla) [inline src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs]: :+1:

**Comment by kichalla** [comment]
Updated /cc: @rynowak @Eilon @HaoK

**Comment by rynowak** [comment]
:shipit:

**Comment by kichalla** [comment]
commit 284c8992333d5eeede4bfd9b34974994606e5c8b

**Pull request closed without merging**